lib/encryptionmanager.cpp

```diff
@@ -0,0 +1,189 @@
+        {
+            Q_ASSERT((0 <= signedKeysProportion) && (signedKeysProportion <= 1));
+            Q_ASSERT((0 <= oneTimeKeyThreshold) && (oneTimeKeyThreshold <= 1));
+            /*
+             * Note about targetKeysNumber:
+             *
+             * From: https://github.com/Zil0/matrix-python-sdk/
+             * File: matrix_client/crypto/olm_device.py
+             *
+             * Try to maintain half the number of one-time keys libolm can hold uploaded
+             * on the HS. This is because some keys will be claimed by peers but not
+             * used instantly, and we want them to stay in libolm, until the limit is reached
+             * and it starts discarding keys, starting by the oldest.
+             */
+        }
+        ~Private()
+        {
+            delete olmAccount;
+        }
+
+        UploadKeysJob* uploadIdentityKeysJob = nullptr;
+        UploadKeysJob* uploadOneTimeKeysJob = nullptr;
+
+        Account* olmAccount;
+        const QByteArray encryptionAccountPickle;
+
+        float signedKeysProportion;
+        float oneTimeKeyThreshold;
+        int targetKeysNumber;
+
+        void updateKeysToUpload();
+        bool oneTimeKeyShouldUpload();
+
+        QHash<QString, int> oneTimeKeyCounts;
+        void setOneTimeKeyCounts(const QHash<QString, int> oneTimeKeyCountsNewValue)
+        {
+            oneTimeKeyCounts = oneTimeKeyCountsNewValue;
+            updateKeysToUpload();
+        }
+        QHash<QString, int> oneTimeKeysToUploadCounts;
+        QHash<QString, int> targetOneTimeKeyCounts
+        {
+            {signedCurve25519Name, qRound(signedKeysProportion * targetKeysNumber)},
+            {curve25519Name, qRound((1-signedKeysProportion) * targetKeysNumber)}
+        };
+};
+
+EncryptionManager::EncryptionManager(const QByteArray &encryptionAccountPickle, float signedKeysProportion, float oneTimeKeyThreshold,
+                                     QObject* parent)
+    : QObject(parent),
+    d(std::make_unique<Private>(std::move(encryptionAccountPickle), std::move(signedKeysProportion), std::move(oneTimeKeyThreshold)))
+{
+
+}
+
+EncryptionManager::~EncryptionManager()
+{
+
+}
+
+void EncryptionManager::uploadIdentityKeys(Connection* connection)
+{
+    DeviceKeys deviceKeys;
+
+    deviceKeys.userId = connection->userId();
+    deviceKeys.deviceId = connection->deviceId();
+
+    QStringList algorithms = { olmCurve25519AesSha256AlgoName,
+                               megolmV1AesShaAlgoName};
+    deviceKeys.algorithms = algorithms;
+
+    QHash<QString, QString> keys;
+    keys.insert(curve25519Name + QStringLiteral(":") + connection->deviceId(),
+                d->olmAccount->curve25519IdentityKey());
+    keys.insert(ed25519Name + QStringLiteral(":") + connection->deviceId(),
+                d->olmAccount->ed25519IdentityKey());
+    deviceKeys.keys = keys;
+
+    QJsonObject message;
+    JsonObjectConverter<DeviceKeys>::dumpTo(message, deviceKeys);
```

Comment:
This is not what I've made the whole `converters.*` thing for :)
`const auto message = toJson(deviceKeys);`
I know the machinery behind that looks downright scary but what you really need to know is that for any serialisable object (i.e. one that has `JsonObjectConverter<>` specialisation) you can use `toJson` (gets data structure, returns JSON), `fillJson` (fills `QJsonObject` in case you already have one), `fromJson` (gets JSON, returns data structure), `fillFromJson` (fills the data structure from JSON).

And, as always, if that doesn't work for whatever reason, let me know and we'll sort it out.